---
title: "User guide: `immunotation` package"
author:
- name: Katharina Imkeller
  affiliation: European Molecular Biology Laboratory, Heidelberg
  email: imkeller@embl.de
output:
  html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(tidyverse)
library(immunotation)
```

## Introduction

### MHC molecules

MHC (major histocompatibility complex) molecules are a family of diverse cell surface complexes that are important for the coordination of adaptive immune responses. MHC molecules are divided into three classes: MHC class I, MHC class II, and non-classical MHC. MHC class I molecules present peptide antigens to CD8+ T cells, whereas MHC class II molecules present epitopes to CD4+ T cells. MHC class I complexes typically consist of one variable alpha chain and one invariant β2-microglobulin chain, whereas MHC class II molecules are composed of one alpha and one beta chain, which are both variable.

MHC (major histocompatibility complex) molecules are cell surface complexes that present peptide antigens to T cells. This processed, also known as MHC-dependent antigen presentation, is essential for the coordination of adaptive immune responses. In humans, MHC molecules are encoded by highly variable HLA (human leukocyte antigen) genes encoded in the hyperpolymorphic HLA locus on chromosome 6. 

MHC restriction ontology

### Hyperpolymorphic HLA genes in human populations

The library of peptide antigens presented on MHC molecules highly depends on the HLA alleles genes present in an individual. The frequencies of individual HLA alleles varies substantially between worldwide human populations.

Reproducible and consistent annotation of HLA alleles in large-scale bioinformatic workflows remains challenging, because the available reference databases and software tools often utilize different HLA naming schemes. This inconsistency mainly results from the different levels of evidence at which HLA allele information is determined (e.g. sequencing protocol or antibody typing).

### Diverse HLA naming schemes
- Protein groups and NMPD MAC (Multiple allele codes)
- old naming schemes
- gene/ allele level

### Variation of HLA alleles across human populations
- AFND database

The Allele Frequency Net Database (AFND, http://www.allelefrequencies.net/) is a repository for immune gene frequencies in different populations worldwide (Gonzalez-Galarza et al. 2020). In addition to a large collection of HLA alle frequency datasets, the database also contains datasets for allele frequencies of KIR (Killer Cell Immunoglobulin-like Receptor) genes, MIC (Major histocompatibility complex class I chain related) and cytokine genes. The current version of AFNDquery allows the automated R access to the HLA related datasets in AFND.

The HLA frequency datasets in AFND are classified according to the following standards:

| Criteria        | Gold standard           | Silver standard  | Bronze standard |
| ----------- |:-----------------:|:-----------------:|:----------:|
| Allele frequency      | Sum to 1 (± 0.015; or ± 0.05) | Sum to 1 (± 0.015; or ± 0.05) | other |
| Sample size     | >= 50 individuals      |   any | any |
| Resolution | four or more digits      | two or more digits | other |

## Scope of the package

The package `immunotation` provides tools for consistent annotation of HLA genes in typical immunoinformatics workflows. 
The package provides functions that allow mapping between different HLA naming schemes and that are based on the MHC restriction ontology (MRO) [cite]. 
The package also provides an automated access to HLA alleles frequencies in worldwide human reference populations stored in the Allele Frequency Net Database [cite].

## Installation

Install from github using the devtools library:

```{r}
#library(devtools)

#devtools::install_github("imkeller/immunotation")
```

## MHC ontology

### Serotypes, genes, alleles in MRO ontology

species

However, very little information is contained for the species 
grass carp, clawed frog, cotton-top tamarin, giant panda, sheep, marmoset
contain little information.

```{r}
get_valid_organisms()
```

numbers

```{r}
df <- retrieve_lookup_table(organism = "human")
```

```{r, warnings = FALSE}
sum_table <- df %>%
    group_by(mhc_type_names, locus_names) %>%
    summarize(number_of_alleles = n())

DT::datatable(sum_table)
```

Serotypes 
```{r}
get_serotypes(allele_list1, mhc_type = "MHC-I")
```
```{r}
get_serotypes(allele_list2, mhc_type = "MHC-II")
```

Sequences

### Functions for mapping between different naming schemes

```{r}
allele_list1 <- c("A*01:01:01", "A*02:01:01",
                "B*39:01:01", "B*07:02:01", 
                "C*08:01:01", "C*01:02:01")
allele_list2 <- c("DPA1*01:03:01", "DPA1*01:04:01",
                  "DPB1*14:01:01", "DPB1*02:01:02",
                "DQA1*02:01:01", "DQA1*05:03",
                "DQB1*02:02:01", "DQB1*06:09:01",
                "DRA*01:01",
                "DRB1*10:01:01", "DRB1*14:02:01")
```

Get protein complexes
```{r}
get_mhcpan_input(allele_list1, mhc_class = "MHC-I")
```

```{r}
get_mhcpan_input(allele_list2, mhc_class = "MHC-II")
```
Get serotypes

### Protein groups, MAC identifiers

# Protein and gene groups

# MACUI

```{r}
allele_list3 <- c("A*01:01:01", "A*02:01:01", "A*03:01")
encode_MAC(allele_list3)
```
```{r}
MAC1 <- "A*01:AYMG"
decode_MAC(MAC1)
```
## Functions related to worldwide population frequencies

### allele frequencies (given allele in several populations)

**Example 1:** Querying the frequency of allele A*02:01 in all populations with more than 10,000 individuals and fulfilling the gold standard.

```{r}
sel1 <- query_allele_frequencies(hla_selection = "A*02:01", 
                                hla_sample_size_pattern = "bigger_than", 
                                hla_sample_size = 10000, 
                                standard="g")

DT::datatable(sel1)
```


**Example 2:** Querying the frequencies of alleles within the HLA-B locus for population "Peru Lamas City Lama" with population_id 1986. If you are not sure which population_id you should use, we recommend to search your population of interest in the *Population* section of the Allele Frequency Net Database and extract the population id from the url (last 4 digits).

```{r}
sel2 <- query_allele_frequencies(hla_locus = "B", hla_population = 1986)

DT::datatable(sel2)
```
```{r}
sel2a <- query_allele_frequencies(hla_locus = "B", hla_population = 3089)

DT::datatable(sel2a)
```

```{r}
plot_allele_frequency(sel1)
```

### comparison of different datasets from same population

### display a geographical location using geographic coordinates

## Querying population metainformation

**Example 3:** Query the metainformation concerning population "Peru Lamas City Lama" (population_id 1986). The webpage concerning the queried information for population Peru Lamas City Lama (1986) can be found here: http://www.allelefrequencies.net/pop6001c.asp?pop_id=1986

```{r}
sel3 <- query_population_detail(1986)

DT::datatable(sel3, options = list(scrollX = TRUE))
```

**Example 4:** Query the metainformation concerning the populations that were listed in the table returned by Example 1

```{r}
sel4 <- query_population_detail(as.numeric(sel1$population_id))

# only select the first 5 columns to discplay in table
DT::datatable(sel4[1:5], options = list(scrollX = TRUE))
```

## References 

1. Gonzalez-Galarza FF, McCabe A, Santos EJ at al. Allele frequency net database (AFND) 2020 update: gold-standard data classification, open access genotype data and new query tools. Nucleic Acid Research (2020).
2. Vita R, Overton JA, Seymour E et al. An ontology for major histocompatibility restriction. J Biomed Semant (2016).
